qos/evm/check.go

```diff
@@ -6,36 +6,39 @@ import (
 	"github.com/buildwithgrove/path/qos/jsonrpc"
 )
 
+// endpointStore provides the endpoint check generator required by
+// the gateway package to augment endpoints' quality data,
+// using synthetic service requests.
+var _ gateway.QoSEndpointCheckGenerator = &endpointStore{}
+
 const (
 	// Each endpoint check should use its own ID to avoid potential conflicts.
 	// ID of JSON-RPC requests for any new checks should be added to the list below.
 	_              = iota
 	idChainIDCheck = 1000 + iota
 	idBlockNumberCheck
-	idArchivalBlockCheck
+	idArchivalCheck
 )
 
-// EndpointStore provides the endpoint check generator required by
-// the gateway package to augment endpoints' quality data,
-// using synthetic service requests.
-var _ gateway.QoSEndpointCheckGenerator = &endpointStore{}
-
-// TODO_IMPROVE(@commoddity): implement QoS check expiry functionality and use protocol.EndpointAddr
-// to filter out checks for any endpoint which has acurrently valid QoS data point.
-func (es *endpointStore) GetRequiredQualityChecks(_ protocol.EndpointAddr) []gateway.RequestQoSContext {
-	qualityChecks := []gateway.RequestQoSContext{
-		getEndpointCheck(es, getChainIDCheckRequest()),
-		getEndpointCheck(es, getBlockNumberCheckRequest()),
-	}
-
-	// If the service is expected to be archival, perform an archival check.
-	if archivalCheckReq, ok := getArchivalCheckRequest(es.serviceState.archivalState); ok {
-		qualityChecks = append(qualityChecks, getEndpointCheck(es, archivalCheckReq))
+// GetRequiredQualityChecks returns the list of quality checks required for an endpoint.
+// It is called in the `gateway/hydrator.go` file on each run of the hydrator.
+func (es *endpointStore) GetRequiredQualityChecks(endpointAddr protocol.EndpointAddr) []gateway.RequestQoSContext {
+	es.endpointsMu.RLock()
+	endpoint, ok := es.endpoints[endpointAddr]
+	es.endpointsMu.RUnlock()
+
+	// If the endpoint is not yet in the store, use an endpoint with the default empty checks.
+	// e.g. if `GetRequiredQualityChecks` is called before the first observation is received for an endpoint.
+	if !ok {
+		endpoint = newEndpoint()
 	}
 
-	return qualityChecks
+	return endpoint.getChecks(es)
```

Comment:
`endpoint` should be a data container only (validation logic may be an acceptable fit). 
Please add the `getEndpointQualityChecks` method to either the endpointStore (it should be part of serviceState eventually, but that seems out of scope for this PR).

Reply:
This makes sense and I'll move it there.